### ISSUE TITLE
AAD Auth (Federated Authenticaion) Support for Azure SQL Database

### ISSUE THREAD

**Issue by gabarach** [dotnet/corefx#8807]
Please add AAD auth support for Azure SQL Database the way how that' s supported in .NET framework 4.6. We should be able to use Authentication modes like "Active Directory Integrated" and "Active Directory Password" in the connecting string.
Currently, including aforementioned authentication modes result into following error:
“Keyword not supported: ‘authentication’ .

**Comment by karelz** [comment]
We need to port over code from full .NET Framework.

**Comment by gabarach** [comment]
So does it mean the AAD auth support for Azure SQL Database may not be available anytime soon?

Thanks & Regards,
Gajendra Barachha

**Comment by karelz** [comment]
Yes, priority-wise it is not too high up given the cost/benefit ratio (as understood today).
If there is significant customer need and feedback, we are always opened to change our point of view. So far this issue has 1 vote.

@YoungGah @saurabh500 are there other ways to authenticate Azure SQL Database from .NET Core?

**Comment by saurabh500** [comment]
Right now connections to Azure SQL DB can be accomplished by specifying the Used Id and Password to the `SqlConnection`

**Comment by karelz** [comment]
@gabarach would that be a reasonable workaround for you?

**Comment by kevcunnane** [comment]
We have a request for the MSSQL extension in VSCode to support Azure AD authentication - see https://github.com/Microsoft/vscode-mssql/issues/435. This is dependent on this enhancement since we use the SqlClient stack in .Net Core to connect. Adding this comment to help track this ask.

**Comment by karelz** [comment]
@kevcunnane why is the workaround above not sufficient for you?

**Comment by kevcunnane** [comment]
@karelz  - the workaround is viable for most cases (in my opinion). I'm trying to add our user's voice to the conversation so that the team can take this on board when prioritizing. I understand this is an enhancement request - for the same reasons that Azure AD support was added to .Net and to ODBC drivers, it would be good to have symmetry here and allow users the benefits that come with Azure AD integration (e.g. no longer manage at a server level, but instead using AD-managed credentials).

As to why I'm raising this: It's something a user of our new MSSQL extension to VSCode raised as an issue. For many users they might not have permissions to create new users and only have an Azure AD login permission, or their company might mandate use of AD authentication. This is lower priority to our team than a number of other .Net Core related issues but it's likely that as Azure AD authentication increases in usage, this may become higher priority.

**Comment by developermj** [comment]
Has there been any traction on this? My employer is moving on-prem applications to the cloud and our current standard is to use Active Directory Integrated authentication. We would really like to use this in our applications. Thanks.

**Comment by movna** [comment]
Any movement on this? Using AAD for Azure SQL is the most secure way available now and having support for this is a must. Thanks.

**Comment by karelz** [comment]
@divega @saurabh500 @corivera can comment on progress / prioritization / workarounds, given we have 10 votes now.
AFAIK we didn't start any work yet.

**Comment by m1nkeh** [comment]
does me giving this issue the thumbs-up count as a "vote" ?

**Comment by saurabh500** [comment]
We are working with the SQL Security team for a AAD auth solution which can work cross platform for .Net Core and which can have a decoupled design for SqlClient in .Net core so that the users who are not using AAD Auth don't need a dependency on all the AAD dependency. 
For AAD Auth on Windows, the design was very tightly coupled with a dependency on an additional DLL written in Native Code which worked on Windows only. The redesign is going going to take some time and collaboration and coordination with multiple internal teams. 

Meanwhile, are folks on this thread looking to using AAD Azure SQL DB auth using only Windows or are there users who intend to take this capability on Linux as well?

**Comment by saurabh500** [comment]
Also what kind of auth satisfy your requirements more? Fedauth username and password or integrated auth?

**Comment by m1nkeh** [comment]
hi @saurabh500 , my requirement is AAD that differs from the log on to my machine (i think this is fedauth ye?) the reason for this is that i work with many clients, and need to log in to multiple Azure AD instances from time to time... so this is my primary requirement.

Regarding platform, i would say that Windows would satisfy my requirement very well, however i am contemplating moving over to work on macOS, and would very much like to be able to use AAD against SQL Svr in Visual Studio Code, but it would be a bonus

**Comment by rgbattersby** [comment]
OSX support would be very well received in my business.

**Comment by saurabh500** [comment]
Thanks @m1nkeh  and @rgbattersby 

@m1nken Yes, I am talking about FedAuth which is used to authenticate to Azure SQL using your domain account.
I want to know if you folks actively use `AAD Integrated Auth` or is it `AAD Username and password` in your connection string? Or is it a good mix of both?

**Comment by m1nkeh** [comment]
AAD Username and Password, yes.

**Comment by rgbattersby** [comment]
Ideally both, to support users who generally belong to other AD environments such as contractors etc and our own employees.

**Comment by mkushakov** [comment]
Since SQL Operations Studio goes to Preview stage, original request of supporting AAD username and password on MacOS and Linux needs to be upvoted and uplifted with higher priority.

**Comment by rlarno** [comment]
Came here from SQL Operation Studio (Preview). As we only allow ad-hoc query to our Azure SQL using Azure AD (username/password), not via SQL Login (username/password).
People are using Windows & MacOS (but mostly Windows)

**Comment by erik-mileiq** [comment]
This is critical for Mac-based teams that are moving to Azure. Please prioritize this.

**Comment by hiramegl** [comment]
AAD Username and Password and windows is what we use in our projects.

**Comment by carbritomcm** [comment]
Came here from SQL Operations Studio (Preview) as well and it would be very important to have AAD authorization working (in both username\password and Integrated).

**Comment by erick-thompson** [comment]
As more and more Azure resources support MSI, we are actively converting all possible resources to use MSI. Why? If there isn't a secret, you can't leak it.

So insofar as this issue relates to MSI support in .NET Core, it is quite important.

**Comment by cleverguy25** [comment]
Is this the right place to up-vote this?  Or is there another place like a user voice site?

**Comment by danielearwicker** [comment]
Workaround is to put the SQL password in a key vault and use MSI to authorise the app to access the key vault, which does work from dotnet core.

**Comment by dougiejay** [comment]
We absolutely want this feature to remove the hassle of keeping passwords in keyvault and then pulling them from .NET core code. The workaround is ok from a security point of view, but adds an extra layer of maintenance that isnt ideal.

**Comment by admin-fsoft** [comment]
Having come from an Azure IaaS approach over to the full PaaS approach with the main goal of reducing administration overhead, this was quite a kick in the balls. 

I also think it is a bit self defeating to not work on this based on up votes as surely the business case there is argument enough without us all having to ask for it (eg, MS pushes integration with Azure AD pretty hard while also telling us how we should all be on .NET Core...). I especially think this is the case as I imagine a certain percentage of users just give up on this approach or accept a degradation of security and manageability rather than going through the process of signing up to GitHub to up vote. We ourselves don't use Github and I created this account solely to up vote it and write this comment.

**Comment by AntonioDrusin** [comment]
We connect to SQL Server (not Azure) from a kubernetes cluster and would love to be able to use AD username and password instead of using SQL authentication. Manageability is key for us.

**Comment by erick-thompson** [comment]
I have a bunch of hocky scripts that I have to use to create users/logins and rotate passwords. It's a lot of fragile code that would go away with integrated security.

**Comment by CarlosOnline** [comment]
Request to add AccessToken to SqlConnection for AAD Service Principle authentication.

`var con = new SqlConnection(builder.ToString()) {AccessToken = token};`

**Comment by Liero** [comment]
@karelz: is 73 votes enough to prioritize this?

The community has become quite vocal about this, especially now when one can easily turn on Managed Service Identity on Azure Web Site

**Comment by karelz** [comment]
@divega @David-Engel @saurabh500 can you please comment on SlqClient plans and how this work fits in?

**Comment by David-Engel** [comment]
We are actively looking in to how and when we can implement this. It seems like the easy first step will be to add AccessToken to SqlConnection as @CarlosOnline mentioned and leave it to the client to obtain the token for now. Following that, we are investigating what we can do on the AAD auth side since we can't use the same underlying libraries for AAD auth that are used in .NET Framework.
David

**Comment by cleverguy25** [comment]
Anything that enables the scenario now is welcome, even if it is more work.  Get a MVP out there and then iterate to improve. Right now we are completely blocked for AAD/Sql Auth.

**Comment by roel4ez** [comment]
@David-Engel a first implementation for which just has the `AccessToken` property on a `SqlConnection` object would already be a big step in the right direction for us. We can let the client obtain the token before passing it to the Connection. 
But right now there is no way for us to securely authenticate a daemon application with a certificate to an Azure SQL database.

**Comment by wokket** [comment]
Another Vote for AccessToken on a .Net Core SqlConnection.  MSI from App Service to Sql Server seems like such a simple no-brainer of a security improvment....

**Comment by perbergland** [comment]
I agree with others: just having accesstoken support would be a major valuable first step and it is currently a big missing piece now that MSI is gaining well worth momentum.

**Comment by iamatinkerer** [comment]
@David-Engel @saurabh500, if the development for AccessToken on SqlConnection hasn't begun, then I could try to implement this using reference source. Let me know what you think.

**Comment by dotnetgator** [comment]
Just to be certain - based on @karelz comments from just a few short months ago (November 2017) - are you certain you understand why AccessToken is being asked for, and why this is not a "number of votes" issue?

**Comment by David-Engel** [comment]
We understand the value of adding the AccessToken property to SqlConnection and how it benefits this issue. There is a separate issue specifically for AccessToken: #13660, It is the next thing we will be working on as soon as currently active work items are wrapped up. @keeratsingh

**Comment by jurjen74** [comment]
This AccessToken is still not available in .NET Core 2.1 / EF Core 2.1.1
Is there an ETA on this?

**Comment by David-Engel** [comment]
@jurjen74 AccessToken support has been added to 2.2 which is planned for Q4 2018: https://github.com/dotnet/corefx/issues/13660#issuecomment-413921458

**Comment by jboarman** [comment]
OK, so now that AccessToken support has been added into 2.2, what steps are needed for full AAD authentication support?

We are moving our teams to Azure MFA authentication, and having this completed is important to completing that migration.  Unless I'm missing something here, applications like SQL Server Operations Studio can't support Azure MFA until AAD Auth support is completed?

**Comment by T16CAG** [comment]
I have been looking on to building a multi tenant application. This would talk to multiple databases. Ideally i do not want to be passing usernames and passwords around. So really would like to have AD integrated logins supported on .NET Core. So count this as an up vote. 

Fairly new to this but would the token approach work in a single application multi tenant environment. Where there is a single application, tenant catalog and multiple contained databases? If anyone could share any links that would be great.

Thank you.

**Comment by schalkje** [comment]
+1 
is a depend for Azure Data Studio MFA logins to Azure DB

**Comment by danmosemsft** [comment]
@David-Engel could you clarify what work remains here, after the 2.2 work mentioned (https://github.com/dotnet/corefx/pull/31039) ? If there's remaining work perhaps it would be helpful to clarify the title.

**Comment by David-Engel** [comment]
@danmosemsft SqlClient in .NET Framework added a SqlAuthenticationProvider interface and implemented an ActiveDirectoryNativeAuthenticationProvider for AAD Integrated authentication. We plan to add the same interface for Core and subsequent implementations so that applications do not have to do the work themselves to obtain access tokens.

**Comment by divega** [comment]
Thank you @danmosemsft and @David-Engel for your help clarifying this. 

So you are saying that what the title asks for "AAD Auth (Federated Authentication) Support for Azure SQL Database" is already supported after https://github.com/dotnet/corefx/issues/13660 got fixed in .NET Core 2.2? Is then fair to say that this issue represents only a different way of doing the same thing but it doesn't really unblock any new scenarios? 

I care a lot about this distinction because I care for the SqlClient team focusing on the really important things, and it seems that the super high vote count on this issue could be the result of a title that is too general, and something like column encryption (https://github.com/dotnet/corefx/issues/18276) is more of a real blocker for customers. 

We can fix the title, but we cannot fix the votes that easily.

**Comment by David-Engel** [comment]
@divega I think that's a fair assessment. Applications will be able to connect to an Azure SQL Database using .NET Core 2.2. They will simply have to implement obtaining the authentication token in the application instead of relying on SqlClient to do it the way it is handled in .NET Framework.

**Comment by divega** [comment]
Thanks @David-Engel. I have tried to capture what this issue is really about in the new title and a note I added on the first comment. Feel free to improve it.

**Comment by ChristopherLClark** [comment]
`dotnet ef dbcontext scaffold` I'm getting the same error keyword is not supported 'authentication`

**Comment by David-Engel** [comment]
@ChristopherLClark Make sure you are using .NET Core 2.2, which is still in preview:
https://www.microsoft.com/net/download/dotnet-core/2.2

**Comment by danmosemsft** [comment]
For 2.2, note that the "Windows  Compatibility Pack" meta package, which is how most people get this, is still pointing to an older SqlClient package. Until that is fixed, you can add a direct reference to the preview SqlClient package: https://www.nuget.org/packages/System.Data.SqlClient/4.6.0-preview3-27014-02

**Comment by danmosemsft** [comment]
cc @safern

**Comment by divega** [comment]
@David-Engel @danmosemsft I think the exception @ChristopherLClark is seeing is expected since SqlClient for .NET Core still does not support using the 'authentication' keyword in the connection string. That will happen when this issue (#8807) is actually fixed. 

In order to use AAD in the new version of SqlClient in .NET Core 2.2, customers need to obtain the access token themselves using ADAL.NET and then set the AccessToken property on the SqlConnection, which was enabled by the fix to https://github.com/dotnet/corefx/pull/31039, as @David-Engel described before.

**Comment by ArmandoLacerda** [comment]
+1

**Comment by AndersMalmgren** [comment]
+1

ETA? Thanks

**Comment by AndersMalmgren** [comment]
In the meantime can we please get sample code how to retrieve Access token correctly

**Comment by zvrba** [comment]
I have the same problem as @AndersMalmgren I tried specifying UserId in the connection string, but then the SqlConnection constructor fails saying that UserId and AccessToken cannot be used together.

**Comment by zvrba** [comment]
@AndersMalmgren I have solved it by using two slashes in the scope: "https://database.windows.net//.default" It is vaguely documented here: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Adal-to-Msal#warning-should-you-have-one-or-two-slashes-in-the-scope-corresponding-to-a-v10-web-api

**Comment by AndersMalmgren** [comment]
Yeah, I got it working too some while ago. 

```csharp
 services.AddScoped<DbContext>(p =>
                    {
                        var ctx = new AuthenticationContext("https://login.microsoftonline.com/xxx");

                        var result = ctx.AcquireTokenAsync("https://database.windows.net/", new ClientCredential("xxx", "xxx"))
                            .ConfigureAwait(false)
                            .GetAwaiter()
                            .GetResult();

                        var db = p.GetService<MyDbContext>();

                        ((SqlConnection)db.Database.GetDbConnection()).AccessToken = result.AccessToken;

                        return db;

                    });
```

**Comment by divega** [comment]
As recently [announced in the .NET Blog](https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/), focus on new SqlClient features an improvements is moving to the new [Microsoft.Data.SqlClient package](https://www.nuget.org/packages/Microsoft.Data.SqlClient/). For this reason, we are moving this issue to the new repo at https://github.com/dotnet/SqlClient. We will still use https://github.com/dotnet/corefx to track issues on other providers like System.Data.Odbc and System.Data.OleDB, and general ADO.NET and .NET data access issues.
